### DISCUSSION TITLE
getting optimal compactness value before running bootstrap

### DISCUSSION THREAD

**Discussion by vanq2024** [JeschkeLab/DeerLab#486] [Q&A]
Back in 2022, Luis had suggested finding the optimal compactness value first before implementing bootstrap to speed things up.

        # Run first run to optimize the compactness via the ICC
        results_ = dl.fit(Vmodel,Vexp, penalties=compactness)
        compactness_weightopt = results_.penweights

       # Run bootstrapped analysis with fixed compactness penalty weight
       compactness.weight.freeze(compactness_weightopt)
       results = dl.fit(Vmodel,Vexp, penalties=compactness, bootstrap=100)

Is this still relevant for the latest DeerLab version (v1.1.4)?
When I tried it, I get errors. Aside from setting bootcores=-1 to use multiple cores, are there other tricks/parameter settings to speed up the fitting beside reducing the bootstrap number?

**Comment by HKaras** [comment]
Hi, 
So I would still recommend fixing the compactness parameter before performing bootstrapping. DeerLab is still relatively general and will rerun all parameters when bootstrapping including the penalities. At the moment it does not make much sense to change this, but maybe we should document it a bit better. 

Further Speedups:
- Use the quadprog package as the `nnlsSolver`. This used to be default but the package is only spoaridcally maintained so is no longer a dependecy. You can install it with `pip install quadprog` and use it with additional fit argument `nnlsSolver='qp'`
- If you are using an AppleSilcon (Mx) processor then you should make sure you use Numpy>2.0. Numpy 2.0 links to the Accelerate backend over OpenBLAS on these processors and is subsequently quicker. 
- If you are using an Intel Processor, then you can try to link numy to Intel-MKL. There are many instructions out there on how to do this, or just ask you favourite AI.

**Comment by vanq2024** [comment]
Hi Hugo. Many thanks for the tips. The first part of Luis's code runs fine. I get error messages for the second part concerning 'list' and 'floats'. What do I need to change in the code?
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[9], line 2
      1 # Run bootstrapped analysis with fixed compactness penalty weight
----> 2 compactness_penalty.weight.freeze(compactness_weightopt)
      3 results = dl.fit(Vmodel,Vexp, penalties=compactness_penalty, bootstrap=20)

File [~\AppData\Local\Programs\Python\Python310\lib\site-packages\deerlab\model.py:188](http://localhost:8888/lab/tree/Juypterlab_EPR/~/AppData/Local/Programs/Python/Python310/lib/site-packages/deerlab/model.py#line=187), in Parameter.freeze(self, value)
    172 def freeze(self,value):
    173     """
    174     Freeze a parameter during a fit/optimization to a given value.
    175 
   (...)
    185         must be within the lower and upper bounds of the parameter.
    186     """
--> 188     if np.any(value>self.ub) or np.any(value<self.lb):
    189         if len(np.atleast_1d(value))>1:
    190             raise ValueError(f"Frozen values are outside of the bounds.")

TypeError: '>' not supported between instances of 'list' and 'float'

**Reply by vanq2024** [reply]
Never mind. A colleague helped me figured it out.
